nemo/collections/tts/modules/audio_codec_modules.py

```diff
@@ -698,3 +716,429 @@ def decode(self, indices: torch.Tensor, input_len: torch.Tensor) -> torch.Tensor
+                    activation=activation,
+                )
+                for dilation in dilations
+            ]
+        )
+
+    def remove_weight_norm(self):
+        for res_block in self.res_blocks:
+            res_block.remove_weight_norm()
+
+    @property
+    def input_types(self):
+        return {
+            "inputs": NeuralType(('B', 'C', 'T'), VoidType()),
+            "input_len": NeuralType(tuple('B'), LengthsType()),
+        }
+
+    @property
+    def output_types(self):
+        return {"out": NeuralType(('B', 'C', 'T'), VoidType())}
+
+    @typecheck()
+    def forward(self, inputs, input_len):
+        out = inputs
+        for res_block in self.res_blocks:
+            out = res_block(inputs=out, input_len=input_len)
+        return out
+
+
+class HiFiGANResLayer(NeuralModule):
+    """
+    Residual block wrapper for HiFi-GAN which creates a block for multiple kernel sizes and dilations.
+    """
+
+    def __init__(self, channels: int, kernel_sizes: Iterable[int], dilations: Iterable[int], activation: str):
+        super().__init__()
+
+        self.res_blocks = nn.ModuleList(
+            [
+                HiFiGANResBlock(channels=channels, kernel_size=kernel_size, dilations=dilations, activation=activation)
+                for kernel_size in kernel_sizes
+            ]
+        )
+
+    def remove_weight_norm(self):
+        for res_block in self.res_blocks:
+            res_block.remove_weight_norm()
+
+    @property
+    def input_types(self):
+        return {
+            "inputs": NeuralType(('B', 'D', 'T'), VoidType()),
+            "input_len": NeuralType(tuple('B'), LengthsType()),
+        }
+
+    @property
+    def output_types(self):
+        return {"out": NeuralType(('B', 'D', 'T'), VoidType())}
+
+    @typecheck()
+    def forward(self, inputs, input_len):
+        residuals = [res_block(inputs=inputs, input_len=input_len) for res_block in self.res_blocks]
+        out = sum(residuals) / len(residuals)
+        return out
+
+
+class HiFiGANDecoder(NeuralModule):
+    """
+    Codec decoder using the HiFi-GAN generator architecture.
+
+    Default parameters match the HiFi-GAN V1 configuration for 22.05khz.
+
+    Args:
+        input_dim: Input dimension.
+        up_sample_rates: Rate to upsample for each decoder block. The product of the upsample rates will
+            determine the output frame rate. For example 8 * 8 * 2 * 2 = 256 samples per token.
+        base_channels: Number of filters in the first convolution. The number of channels will be cut in
+            half after each upsample layer.
+        in_kernel_size: Kernel size of the input convolution.
+        out_kernel_size: Kernel size of the output convolution.
+        resblock_kernel_sizes: List of kernel sizes to use in each residual block.
+        resblock_dilation_sizes: List of dilations to use in each residual block.
+        activation: Activation to use, defaults to leaky relu.
```

Comment:
May be write: Activation in upsample layers? As Tanh is used for final activation.

Reply:
Changed.